mixnet/sim/stats.py

```diff
@@ -0,0 +1,115 @@
+from collections import Counter, defaultdict
+
+import matplotlib.pyplot as plt
+import numpy
+import pandas
+from matplotlib.axes import Axes
+
+from protocol.node import Node
+from sim.connection import MeteredRemoteSimplexConnection
+
+# A map of nodes to their inbound/outbound connections
+NodeConnectionsMap = dict[
+    Node,
+    tuple[list[MeteredRemoteSimplexConnection], list[MeteredRemoteSimplexConnection]],
+]
+
+
+class ConnectionStats:
+    def __init__(self):
+        self.conns_per_node: NodeConnectionsMap = defaultdict(lambda: ([], []))
+
+    def register(
+        self,
+        node: Node,
+        inbound_conn: MeteredRemoteSimplexConnection,
+        outbound_conn: MeteredRemoteSimplexConnection,
+    ):
+        self.conns_per_node[node][0].append(inbound_conn)
+        self.conns_per_node[node][1].append(outbound_conn)
+
+    def analyze(self):
+        self.__message_sizes()
+        self.__bandwidths_per_conn()
+        self.__bandwidths_per_node()
+
+    def __message_sizes(self):
+        """
+        Analyzes all message sizes sent across all connections of all nodes.
+        """
+        sizes: Counter[int] = Counter()
+        for _, (_, outbound_conns) in self.conns_per_node.items():
+            for conn in outbound_conns:
+                sizes.update(conn.msg_sizes)
+
+        df = pandas.DataFrame.from_dict(sizes, orient="index").reset_index()
+        df.columns = ["msg_size", "count"]
+        print("==========================================")
+        print("Message Size Distribution")
+        print("==========================================")
+        print(f"{df}\n")
+
+    def __bandwidths_per_conn(self):
+        """
+        Analyzes the bandwidth consumed by each simplex connection.
+        """
+        plt.plot(figsize=(12, 6))
```

Comment:
Is the size fixed for a reason? Shouldn't it be somehow calculated dynamically, for example, based on the size of the network or peering degree?

Reply:
That's a good idea. I used the fixed size because I was not familiar with matplotlib. Let me check if there is any dynamic way. I will open a new PR for that.